library/alloc/src/vec.rs

```diff
@@ -2391,11 +2384,24 @@ where
 }
 
 impl<T> SpecExtend<T, IntoIter<T>> for Vec<T> {
-    fn spec_extend(&mut self, mut iterator: IntoIter<T>) {
-        unsafe {
-            self.append_elements(iterator.as_slice() as _);
+    fn spec_extend(&mut self, iterator: IntoIter<T>) {
+        // Avoid reallocation if we can use iterator's storage instead. This requires 1 memcpy and 0-1 memmove
+        // while reallocation would require 1 alloc, 1-2 memcpy, 1-2 free
+        if mem::size_of::<T>() > 0
+            && self.capacity() - self.len() < iterator.len()
+            && iterator.cap - iterator.len() >= self.len()
+        {
+            // Safety: we just checked that IntoIter has sufficient capacity to prepend our elements.
+            // Prepending will then fill the uninitialized prefix.
+            *self = unsafe {
+                let mut v = iterator.into_vec(self.len() as isize);
```

Comment:
`iterator.into_vec` is doing a `copy` but here we don't need it, we would rather leave the data uninitialized to reduce `memcpy`.

Reply:
`into_vec` shifts to the right to create the gap into which we copy. we need both.

Reply:
Okay, I was confused for a moment. I forgot `into_vec` was taking in an `offset` rather than `len`, maybe we should call it `into_vec_with_offset`?

Reply:
Renamed to `into_vec_with_uninit_prefix` to highlight at least part of the unsafety.